llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -3145,12 +3153,16 @@ InstructionCost AArch64TTIImpl::getCFInstrCost(unsigned Opcode,
   return 0;
 }
 
-InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(const Instruction *I,
-                                                         Type *Val,
-                                                         unsigned Index,
-                                                         bool HasRealUse) {
+InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
+    std::variant<const Instruction *, const unsigned> InstOrOpcode, Type *Val,
```

Comment:
1. Sometimes Instruction itself is passed to this API to get more information. So, `const Instruction *I` has to be one of the arguments
2. The thinking behind using single variable for Instruction and Opcode(using variant) is Instruction and Opcode is somewhat related but we never pass both unless its required in future.

Reply:
You really cant merge the code for instruction and opcode, atleast for the patch, because the change wont be reflected in the cost-model tests.  
1. When `opcode!=null` , less information is available  and we bail out early in general. 
2. When `I!=null`, more analysis needs to be done.

I will keep seperate `instruction` and `opcode` parameter as you say.